hydroflow/src/builder/build/pull_batch.rs

```diff
@@ -42,42 +43,37 @@ where
             prev_a,
             prev_b,
             state: Default::default(),
+            _marker: PhantomData,
         }
     }
 }
 
-impl<PrevBuf, PrevStream, Key, BufVal, StreamVal> PullBuildBase
-    for BatchPullBuild<PrevBuf, PrevStream, Key, BufVal, StreamVal>
+impl<PrevBuf, PrevStream, L, Update, Tick> PullBuildBase
+    for BatchPullBuild<PrevBuf, PrevStream, L, Update, Tick>
 where
-    PrevBuf: PullBuild<ItemOut = (Key, BufVal)>,
-    PrevStream: PullBuild<ItemOut = (Key, StreamVal)>,
-    Key: 'static + Eq + Hash,
-    BufVal: 'static,
-    StreamVal: 'static,
+    PrevBuf: PullBuild<ItemOut = Update::Repr>,
+    PrevStream: PullBuild<ItemOut = Tick>,
+    Update: 'static + LatticeRepr,
+    L: 'static + LatticeRepr + Merge<Update>,
+    L::Repr: Default,
 
     PrevBuf::InputHandoffs: Extend<PrevStream::InputHandoffs>,
     <PrevBuf::InputHandoffs as Extend<PrevStream::InputHandoffs>>::Extended: PortList<RECV>
         + PortListSplit<RECV, PrevBuf::InputHandoffs, Suffix = PrevStream::InputHandoffs>,
 {
-    type ItemOut = (Key, StreamVal, Vec<BufVal>);
-    type Build<'slf, 'hof> = BatchJoin<
-        'slf,
-        Key,
-        PrevBuf::Build<'slf, 'hof>,
-        BufVal,
-        PrevStream::Build<'slf, 'hof>,
-        StreamVal,
-    >;
+    type ItemOut = (Tick, L::Repr);
```

Comment:
You could get rid of the `Tick` generic param if you wanted to by using `(PrevStream::ItemOut, L::Repr)` I think, but either way is good.